### ISSUE TITLE
regression involving subclass and overloaded constructor

### ISSUE THREAD

**Issue by cleoold** [microsoft/pyright#3927]
**Describe the bug**
Following code reports error. If we move the `method` to the parent class the error disappears. If we explicitly annotate the `inner()` with ` -> Iterable[ContainerSub[T_co]]` instead of a `generator` the error also disappears. Replacing the `__init__`'s arg with `Union` instead of overloading can also avoid the error.

**To Reproduce**
Steps to reproduce the behavior.

**Expected behavior**
Code used to type check before 1.1.262

**Screenshots or Code**
```python
T_co = TypeVar('T_co', covariant=True)

class Container(Generic[T_co]):
    @overload
    def __init__(self, arg: Iterable[T_co]): ...
    @overload
    def __init__(self, arg: Callable[[], Iterable[T_co]]): ...
    def __iter__(self) -> Iterator[T_co]: ...


class ContainerSub(Container[T_co]):
    def method(self) -> 'ContainerSub[ContainerSub[T_co]]':
        def inner():  # Generator[ContainerSub[T_co@ContainerSub], None, None]
            for x in self:
                yield ContainerSub(lambda: [x])
        return ContainerSub(inner)
                   ^~~~~~~~~~~~~
error: Expression of type "ContainerSub[ContainerSub[ContainerSub[T_co@ContainerSub]]]" cannot be assigned to return type "ContainerSub[ContainerSub[T_co@ContainerSub]]"
    TypeVar "T_co@ContainerSub" is covariant
      TypeVar "T_co@ContainerSub" is covariant
        Type "ContainerSub[T_co@ContainerSub]" cannot be assigned to type "T_co@ContainerSub" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
pyright 1.1.270

**Additional context**

**Comment by erictraut** [comment]
Thanks for the bug report. This was due to an existing bug that was being masked previously. A change related to another bug fix exposed this one.

I'll need to investigate how to fix this without breaking other cases. In the meantime, you can work around this issue by providing an explicit return type annotation for `inner`.

```python
        def inner() -> Generator[ContainerSub[T_co], None, None]:
            for x in self:
                yield ContainerSub(lambda: [x])
```

**Comment by cleoold** [comment]
> Thanks for the bug report. This was due to an existing bug that was being masked previously. A change related to another bug fix exposed this one.
> 
> I'll need to investigate how to fix this without breaking other cases. In the meantime, you can work around this issue by providing an explicit return type annotation for `inner`.
> 
> ```python
>         def inner() -> Generator[ContainerSub[T_co], None, None]:
>             for x in self:
>                 yield ContainerSub(lambda: [x])
> ```

thanks. I didn't know annotating `Generator` would also work

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.272. It will also be included in a future release of pylance.
